xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/test/java/org/xwiki/rest/internal/ModelFactoryTest.java

```diff
@@ -0,0 +1,142 @@
+        
+        BaseObject xwikiObject = mock(BaseObject.class);
+        BaseClass xwikiClass  = mock(BaseClass.class);
+       
+        when(xwikiObject.getPropertyNames()).thenReturn(new String[] {});
+        when(xwikiObject.getXClass(xcontext)).thenReturn(xwikiClass);
+        when(xwikiObject.getClassName()).thenReturn("Some.XClass");
+        when(xwikiObject.getNumber()).thenReturn(0);
+        
+        StringClass stringField = new StringClass();
+        stringField.setName("textValue");
+        StringProperty textElement = new StringProperty();
+        textElement.setName("textValue");
+        textElement.setClassType("String");
+        textElement.setValue("abcd");
+        when(xwikiObject.get("textValue")).thenReturn(textElement);
+
+        PasswordClass pwField = new PasswordClass();
+        pwField.setName("passwordValue");
+        StringProperty pwElement = new StringProperty();
+        pwElement.setName("passwordValue");
+        pwElement.setClassType("Password");
+        pwElement.setValue("abcd");
+        when(xwikiObject.get("passwordValue")).thenReturn(pwElement);
+        
+        when(xwikiClass.getProperties()).thenReturn(new java.lang.Object[] {stringField, pwField});
+        
+        // some more mockery needed to pass the "toRestObject" call
+        // although the results are not tested (yet)
+
+        URI dummy = new URI("https://localhost/");
+        Document doc = mock(Document.class);
+        SpaceReference spaceRef = new SpaceReference("Space", new WikiReference("wiki"));
+        when(doc.getPrefixedFullName()).thenReturn("wiki:Space.Page");
+        when(doc.getWiki()).thenReturn("wiki");
+        when(doc.getSpace()).thenReturn("Space");
+        when(doc.getName()).thenReturn("Page");
+        when(doc.getDocumentReference()).thenReturn(new DocumentReference("Page", spaceRef));
+        when(resolver.resolve(eq("Space"), any(String.class))).thenReturn(spaceRef);
+        
+        com.xpn.xwiki.web.Utils.setComponentManager(componentManager);
+        
+        //
+        // now we get around to call the method under test
+        //
+        
+        Object result = modelFactory.toRestObject(dummy, doc, xwikiObject, false, false);
+
+        List<Property> properties = result.getProperties();
+        // note: we do not mind if there is no "Password" property at all in the list
+        // that is as good as not showing the value
+        boolean stringFieldFound = false;
+        for (Property prop : properties) {
+            if ("String".equals(prop.getType())) {
+                assertEquals("abcd", prop.getValue(), "should see proper value for StringField");
+                stringFieldFound = true;
+            }
+            if ("Password".equals(prop.getType())) {
+                assertNull(prop.getValue(), "should see not value for PasswordField");
+            }
+        }
+        assertTrue(stringFieldFound, "should have StringField in the result");
```

Comment:
Thanks @ClemensRobbenhaar for writing a test. Does it mean we don't have one that already calls REST and check the result?

Reply:
From What I see at PageResourceTest, ideally we should add a new test method there like `testPUTGETWithPasswordXProperty` or modify the existing `testPUTGETWithObject` but instead of using the Tag XObject, use a XWikiUsers xobjects for example (which contains a password field).

You test verifies that we don't show the password which is cool but it doesn't test that we can set it. Doing it the way it's done in PageResourceTest would validate both aspects (PUT and GET).

Note to self: we need to convert PageResourceTest to a JUnit5 test.

@ClemensRobbenhaar Don't mistake me. I'm very happy that you wrote the fix + the test. For me that's good enough to close the issue.

Reply:
I agree that the test is not perfect and there is much to be improved. Especially I know the "set password" is not tested. Today I wondered if there is a use case for it; i.e. creating users with a password via REST. If the setter for the password accepts the unencrypted password and encrypts it, this might even be working, and it is probably not a security issue, as access rights are checked by the REST API. So maybe I broke something here instead if fixing it.

I did not notice that ``PageResourceTest`` would be the place to extend an existing test, or sure I would have been gone that way instead.

Reply:
Ah, ``PageResourceTest`` is a HTTP-Test. I did not look there ... these integration tests are quite a different thing to write and run, at least for me.
The unit test has been improved according to your comments.
 
If password fields should still be writeable, I have to undo a change in the implementation (remove the check in ``ModelFactory#toObject``)